app/services/puppet_fact_parser.rb

```diff
@@ -190,48 +197,127 @@ def use_legacy_facts?
         '11'
       else
         release
       end
     else
-      facts.dig(:os, :release, :full) || facts[:lsbdistrelease] || facts[:operatingsystemrelease]
+      os_release_full
     end
   end
+
+  # The full OS release (7 / 7.9 / 7.6.1810 / 2012 R2 / 20.04)
+  def os_release_full
+    # Facter 2.2 introduced the os.release fact
+    facts.dig(:os, :release, :full) || facts[:operatingsystemrelease]
+  rescue StandardError => e
+    logger.error { "Failed to read the full OS release: #{e}" }
+    nil
+  end
+
+  # This fact returns the distribution's id, which typically relies on
+  # lsb-release to be installed. As such, it's an optional fact
+  def distro_id
+    # Facter 3.0 introduced the os.distro fact
+    facts.dig(:os, :distro, :id).presence || facts[:lsbdistid].presence
+  rescue StandardError => e
+    logger.warning { "Failed to the read distribution id: #{e}" }
+    nil
+  end
+
+  # This fact returns the distribution's codename, which typically relies on
+  # lsb-release to be installed. As such, it's an optional fact
+  def distro_codename
+    # Facter 3.0 introduced the os.distro fact
+    facts.dig(:os, :distro, :codename).presence || facts[:lsbdistcodename].presence
+  rescue StandardError => e
+    logger.warning { "Failed to read the distribution codename: #{e}" }
+    nil
+  end
+
+  # This fact returns the distribution's description, which typically relies on
+  # lsb-release to be installed. As such, it's an optional fact
+  def distro_description
+    # Facter 3.0 introduced the os.distro fact
+    facts.dig(:os, :distro, :description).presence || facts[:lsbdistdescription].presence
+  rescue StandardError => e
+    logger.warning { "Failed to read the distribution description: #{e}" }
+    nil
+  end
+
+  # Product name from DMI
+  def dmi_product_name
+    # Facter 3.0 introduced the dmi fact
+    facts.dig(:dmi, :product, :name).presence || facts[:productname]
+  rescue StandardError => e
+    logger.warning { "Failed to read the product name: #{e}" }
+    nil
+  end
+
+  # Board product name as the DMI board reports it.
+  def dmi_board_product
+    # Facter 3.0 introduced the dmi fact
+    facts.dig(:dmi, :board, :product).presence || facts[:boardproductname]
+  rescue StandardError => e
+    logger.warning { "Failed to read the board product: #{e}" }
+    nil
+  end
+
+  # Architecture (x86_64 / amd64 /x64 / i386 / x64).
+  def architecture_fact
+    # Facter 3.0 introduced the os.architecture fact
+    facts.dig(:os, :architecture).presence || facts[:architecture].presence
+  rescue StandardError => e
+    logger.error { "Failed to read the architecture: #{e}" }
+    nil
+  end
+
+  # Hardware ISA (x86_64 / i686 / i386).
+  def hardware_isa
+    # Facter 3.0 introduced the processors.isa fact
+    facts.dig(:processors, :isa).presence || facts[:hardwareisa].presence
+  rescue StandardError => e
+    logger.error { "Failed to read the hardware ISA: #{e}" }
+    nil
+  end
```

Comment:
Please move all those methods into the public section, there are already a few and these can be pretty useful down the road.

Reply:
The reason for this is I want you to write unit tests for all these methods.